src/common/ad_tx.c

```diff
@@ -608,3 +613,423 @@ ad_tx_redo_act_next(struct ad_tx *tx)
+	 * the explicit umem_free() on error cleanup is necessary for
+	 * non-transactional case.
+	 */
+	if (ad_tx_stage(tx) == UMEM_STAGE_ONABORT)
+		return 0;
+
+	if (!UMOFF_IS_NULL(umoff))
+		rc = ad_tx_free(tx, umoff);
+
+	return rc;
+}
+
+umem_off_t
+mo_ad_tx_alloc(struct umem_instance *umm, size_t size, int slab_id, uint64_t flags,
+	       unsigned int type_num)
+{
+	struct ad_blob_handle	 bh = umm2ad_blob_hdl(umm);
+
+	return ad_alloc(bh, 0, size, NULL);
+}
+
+int
+mo_ad_tx_add(struct umem_instance *umm, umem_off_t umoff, uint64_t offset, size_t size)
+{
+	struct ad_tx		*tx = tx_get();
+	struct ad_blob_handle	 bh = umm2ad_blob_hdl(umm);
+	struct ad_blob		*blob = bh.bh_blob;
+	void			*ptr;
+
+	D_ASSERT(offset == 0);
+	D_ASSERTF(ad_tx_stage(tx) == UMEM_STAGE_WORK,
+		  "TX "DF_U64", bad stage %d\n", ad_tx_id(tx), ad_tx_stage(tx));
+	ptr = blob_addr2ptr(blob, umoff);
+	return ad_tx_snap(tx, ptr, size, AD_TX_UNDO);
+}
+
+int
+mo_ad_tx_xadd(struct umem_instance *umm, umem_off_t umoff, uint64_t offset, size_t size,
+	      uint64_t flags)
+{
+	/* NOOP for UMEM_XADD_NO_SNAPSHOT */
+	if (flags & UMEM_XADD_NO_SNAPSHOT)
+		return 0;
+
+	return mo_ad_tx_add(umm, umoff, offset, size);
+}
+
+int
+mo_ad_tx_add_ptr(struct umem_instance *umm, void *ptr, size_t size)
+{
+	struct ad_tx		*tx = tx_get();
+
+	D_ASSERTF(ad_tx_stage(tx) == UMEM_STAGE_WORK,
+		  "TX "DF_U64", bad stage %d\n", ad_tx_id(tx), ad_tx_stage(tx));
+
+	return ad_tx_snap(tx, ptr, size, AD_TX_UNDO);
+}
+
+umem_off_t
+mo_ad_reserve(struct umem_instance *umm, struct pobj_action *pact, size_t size,
+	      unsigned int type_num)
+{
+	struct ad_reserv_act	*act = (struct ad_reserv_act *)pact;
+	struct ad_blob_handle	 bh = umm2ad_blob_hdl(umm);
+
+	return ad_reserve(bh, 0, size, NULL, act);
+}
+
+void
+mo_ad_cancel(struct umem_instance *umm, struct pobj_action *actv, int actv_cnt)
+{
+	struct ad_reserv_act	*act;
+	int			 i;
+
+	for (i = 0; i < actv_cnt; i++) {
+		act = (struct ad_reserv_act *)(&actv[i]);
+		ad_cancel(act, 1);
+	}
+}
+
+int
+mo_ad_tx_publish(struct umem_instance *umm, struct pobj_action *actv, int actv_cnt)
+{
+	struct ad_tx		*tx = tx_get();
+	struct ad_reserv_act	*act;
+	int			 i;
+	int			 rc = 0;
+
+	D_ASSERTF(ad_tx_stage(tx) == UMEM_STAGE_WORK,
+		  "TX "DF_U64", bad stage %d\n", ad_tx_id(tx), ad_tx_stage(tx));
+
+	for (i = 0; i < actv_cnt; i++) {
+		act = (struct ad_reserv_act *)(&actv[i]);
+		rc = ad_tx_publish(tx, act, 1);
+		if (rc) {
+			D_ERROR("TX "DF_U64", ad_tx_publish failed, "DF_RC"\n",
+				ad_tx_id(tx), DP_RC(rc));
+			break;
+		}
+	}
+
+	return rc;
+}
+
+void *
+mo_ad_atomic_copy(struct umem_instance *umm, void *dest, const void *src, size_t len)
+{
+	struct ad_tx	*tx = tx_get();
+	bool		 tx_started = false;
+	int		 rc = 0;
+
+	if (tx == NULL) {
+		rc = mo_ad_tx_begin(umm, NULL);
+		if (rc) {
+			D_ERROR("mo_ad_tx_begin failed, "DF_RC"\n", DP_RC(rc));
+			return NULL;
+		}
+		tx_started = true;
+		tx = tx_get();
+	}
```

Comment:
Simply call tx_begin() then tx_commit() should be fine since we support nested transaction, but I think there won't be any atomic copy use case for md on SSD anyway.

Reply:
I think Niu suggests you to call mo_ad_tx_begin anyway, instead of doing those checks, because we can support nested transaction.